File: 행렬동적할당/matrix.c
#include <stdio.h>
#include <stdlib.h>

int** allocateMatrix(int rows, int cols) {
    int** matrix = (int**)malloc(rows * sizeof(int*));
    for (int i = 0; i < rows; i++) {
        matrix[i] = (int*)malloc(cols * sizeof(int));
    }
    return matrix;
}

void deallocateMatrix(int** matrix, int rows) {
    for (int i = 0; i < rows; i++) {
        free(matrix[i]);
    }
    free(matrix);
}

void inputMatrix(int** matrix, int rows, int cols) {
    for (int i = 0; i < rows; i++) {
	  for (int j = 0; j < cols; j++) {
               printf("Enter element [%d][%d]: ", i, j);
               scanf("%d", &matrix[i][j]);
            }
       }
}

void addMatrices(int** matrix1, int** matrix2, int** result, int rows, int cols) {
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            result[i][j] = matrix1[i][j] + matrix2[i][j];
	        }
        }
    }

void printMatrix(int** matrix, int rows, int cols) {
    for (int i = 0; i < rows; i++) {
         for (int j = 0; j < cols; j++) {
              printf("%d ", matrix[i][j]);
        }
        printf("\n");
     }
}

int main() {
    int rows, cols;
    
    printf("Enter the number of rows: ");
    scanf("%d", &rows);

    printf("Enter the number of columns: ");
    scanf("%d", &cols);

    int** matrix1 = allocateMatrix(rows, cols);
    int** matrix2 = allocateMatrix(rows, cols);
    int** result = allocateMatrix(rows, cols);



    printf("Enter elements for matrix 1:\n");
    inputMatrix(matrix1, rows, cols);


    printf("Enter elements for matrix 2:\n");
    inputMatrix(matrix2, rows, cols);

    addMatrices(matrix1, matrix2, result, rows, cols);

    printf("Resultant matrix after addition:\n");
    printMatrix(result, rows, cols);


    deallocateMatrix(matrix1, rows);
    deallocateMatrix(matrix2, rows);
    deallocateMatrix(result, rows);

    return 0;
}

File: 행렬동적할당/matrix.rs
fn add_matrices(matrix1: &Vec<Vec<i32>>, matrix2: &Vec<Vec<i32>>) -> Vec<Vec<i32>> {
    let rows = matrix1.len();
    let cols = matrix1[0].len();

    let mut result = vec![vec![0; cols]; rows];

    for i in 0..rows {
        for j in 0..cols {
            result[i][j] = matrix1[i][j] + matrix2[i][j];                 
        }
    }
   
    result
}

fn main() {
    let mut input = String::new();
    println!("Enter the number of rows: ");
    std::io::stdin().read_line(&mut input).unwrap();
    let rows: usize = input.trim().parse().unwrap();


    
    input.clear();
    println!("Enter the number of columns: ");
    std::io::stdin().read_line(&mut input).unwrap();
    let cols: usize = input.trim().parse().unwrap();


    
    let mut matrix1 = vec![vec![0; cols]; rows];
    let mut matrix2 = vec![vec![0; cols]; rows];


    
    println!("Enter elements for matrix 1:");
    for i in 0..rows {
        for j in 0..cols {
            input.clear();
            std::io::stdin().read_line(&mut input).unwrap();
            matrix1[i][j] = input.trim().parse().unwrap();   
        }
    }


    
    println!("Enter elements for matrix 2:");
    for i in 0..rows {
        for j in 0..cols {
            input.clear();
            std::io::stdin().read_line(&mut input).unwrap();
            matrix2[i][j] = input.trim().parse().unwrap();   
        }                                              
    }


    
    let result = add_matrices(&matrix1, &matrix2);
    
    println!("Resultant matrix after addition:");
    for i in 0..rows {
        for j in 0..cols {
            print!("{} ", result[i][j]);
            
        }
        println!();
    }
}
